homeassistant/components/tuya/alarm_control_panel.py

```diff
@@ -0,0 +1,149 @@
+"""Support for Tuya Alarm."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+import logging
+
+from tuya_iot import TuyaDevice, TuyaDeviceManager
+
+from homeassistant.components.alarm_control_panel import (
+    AlarmControlPanelEntity,
+    AlarmControlPanelEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import HomeAssistantTuyaData
+from .base import TuyaEntity
+from .const import DOMAIN, TUYA_DISCOVERY_NEW, DPCode
+
+_LOGGER = logging.getLogger(__name__)
+
+ICON = "mdi:security"
+
+SUPPORT_ALARM_ARM_HOME = 1
+SUPPORT_ALARM_ARM_AWAY = 2
+SUPPORT_ALARM_ARM_NIGHT = 4
+SUPPORT_ALARM_TRIGGER = 8
+SUPPORT_ALARM_ARM_CUSTOM_BYPASS = 16
+SUPPORT_ALARM_ARM_VACATION = 32
+
+
+@dataclass
+class TuyaAlarmEntityDescription(AlarmControlPanelEntityDescription):
+    """Describe an Tuya cover entity."""
+
+    battery_percentage: DPCode | None = None
+
+
+# All descriptions can be found here:
+# https://developer.tuya.com/en/docs/iot/standarddescription?id=K9i5ql6waswzq
+ALARM: dict[str, tuple[TuyaAlarmEntityDescription, ...]] = {
+    # Alarm
+    # https://developer.tuya.com/en/docs/iot/categorymal?id=Kaiuz33clqxaf
+    "mal": (
+        TuyaAlarmEntityDescription(
+            key="master_mode",
+            name="Alarm",
+            battery_percentage=DPCode.BATTERY_PERCENTAGE,
+        ),
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Tuya alarm dynamically through Tuya discovery."""
+    hass_data: HomeAssistantTuyaData = hass.data[DOMAIN][entry.entry_id]
+    _LOGGER.info(hass_data)
+
+    @callback
+    def async_discover_device(device_ids: list[str]) -> None:
+        """Discover and add a discovered Tuya siren."""
+        entities: list[TuyaAlarmEntity] = []
+        for device_id in device_ids:
+            device = hass_data.device_manager.device_map[device_id]
+            if descriptions := ALARM.get(device.category):
+                for description in descriptions:
+                    if (
+                        description.key in device.function
+                        or description.key in device.status
```

Comment:
The key checked, is not the key used in the entity. So either this check isn't correct, or the description key isn't correct.

Reply:
@frenck  i'm not sure but this is the value of device.status:

`{'master_mode': 'arm', 'delay_set': 0, 'alarm_time': 3, 'switch_alarm_sound': True, 'switch_alarm_light': False, 'switch_kb_sound': False, 'switch_kb_light': True, 'muffling': False, 'alarm_msg': 'xxx', 'switch_alarm_propel': False, 'alarm_delay_time': 20, 'master_state': 'normal', 'sub_class': 'remote_controller', 'sub_admin': 'Bv8=', 'sub_state': 'normal'}
`

and description.key should be

`master_mode`

Reply:
Sure, but it isn't used by the entity. So, so something is wrong with this logic.

Reply:
i suppose i see, i've use the wrong DPCode naming. I've named them as alarm_state instead of alarm_mode